### PR TITLE
CI: Add workflow to build and upload free-threaded wheels

### PR DESCRIPTION
**Pull request by andfoy**
#### Reference issue

See #20669 

#### What does this implement/fix?

This PR adds a GitHub actions workflow to release free-threaded nightly wheels on the Scientific Python Anaconda wheel channel

#### Additional information

Corresponding PR in NumPy: https://github.com/numpy/numpy/pull/26512

### CODE REVIEW

File: .github/workflows/free_threaded_wheels.yml [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    runs-on: ${{ matrix.buildplat[0] }}
+    strategy:
+      # Ensure that a wheel builder finishes even if another fails
+      fail-fast: false
+      matrix:
+        # Github Actions doesn't support pairing matrix values together, let's improvise
+        # https://github.com/github/feedback/discussions/7835#discussioncomment-1769026
+        buildplat:
+          - [ubuntu-20.04, manylinux_x86_64, ""]
+          - [ubuntu-20.04, musllinux_x86_64, ""]
```
Comment (rgommers) [inline .github/workflows/free_threaded_wheels.yml:84]: Why isn't this 22.04 like in `wheels.yml`?

Comment (rgommers) [reply .github/workflows/free_threaded_wheels.yml:84]: It's also missing a set of quotation marks, which will make `matrix.buildplat[4]` go missing.

File: tools/wheels/cibw_before_build_linux.sh [hunk 2]
```diff
@@ -17,6 +17,18 @@ printenv
...
 
+# TODO: delete along with enabling build isolation by unsetting
+# CIBW_BUILD_FRONTEND when scipy is buildable under free-threaded
+# python with a released version of cython
+FREE_THREADED_BUILD="$(python -c"import sysconfig; print(bool(sysconfig.get_config_var('Py_GIL_DISABLED')))")"
+if [[ $FREE_THREADED_BUILD == "True" ]]; then
+    python -m pip install -U --pre pip
+    python -m pip install git+https://github.com/cython/cython
+    python -m pip install -i https://pypi.anaconda.org/scientific-python-nightly-wheels/simple numpy
+    python -m pip install git+https://github.com/serge-sans-paille/pythran
```
Comment (rgommers) [inline tools/wheels/cibw_before_build_linux.sh:28]: Are there any known issues in `pythran`? If not, easier to add `pythran` to the next line next to `pybind11`.

File: .github/workflows/free_threaded_wheels.yml [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+            LD_LIBRARY_PATH=$LIB_PATH:$LD_LIBRARY_PATH\
+            SDKROOT=$(xcrun --sdk macosx --show-sdk-path)\
+            PIP_PRE=1\
+            PIP_NO_BUILD_ISOLATION=false\
+            PKG_CONFIG_PATH=$LIB_PATH/pkgconfig\
+            PIP_EXTRA_INDEX_URL=https://pypi.anaconda.org/scientific-python-nightly-wheels/simple"
+          echo "CIBW_ENVIRONMENT_MACOS=$CIBW" >> "$GITHUB_ENV"
+
+          echo "REPAIR_PATH=$LIB_PATH" >> "$GITHUB_ENV"
+          GFORTRAN_LIB="\$(dirname \$(gfortran --print-file-name libgfortran.dylib))"
```
Comment (rgommers) [inline .github/workflows/free_threaded_wheels.yml:151]: Note that this part just under surgery at gh-20870, so let's do the pragmatic thing here and remove everything related to macOS and Windows from this file.

**Review by rgommers — CHANGES_REQUESTED** [review]
Thanks @andfoy. A few comments inline.

Once this works on your fork, it needs a push to this PR with `[wheel build]` on the first line of the newest commit message, to trigger a successful run of the new job here.

File: .github/workflows/free_threaded_wheels.yml (hunk 1 above)
Comment (andfoy) [reply .github/workflows/free_threaded_wheels.yml:84]: I copied this part over from NumPy PR, I can sync it with the values at `wheels.yml`

**Review by rgommers — APPROVED** [review]
This looks good to me now, thanks @andfoy.

The build log is clean and things seem to work as expected. There is one relevant warning:
```
  + export PYTHON_GIL=0
  + PYTHON_GIL=0
...

   + python -c 'import sys; import scipy; sys.exit(not scipy.test())'
  <frozen importlib._bootstrap>:488: RuntimeWarning: The global interpreter lock (GIL) has been enabled to load module 'numpy._core._multiarray_umath', which has not declared that it can run safely without the GIL. To override this behavior and keep the GIL disabled (at your own risk), run with PYTHON_GIL=0 or -Xgil=0.
```
however that is either a numpy or a Python issue; it's visible in the numpy free-threaded wheel build [CI log](https://github.com/numpy/numpy/actions/runs/9377383356/job/25818853571) as well. @ngoldbaum is this a known thing?

**Comment by rgommers** [comment]
manylinux/musllinux `cp313t` nightly wheels are up at https://anaconda.org/scientific-python-nightly-wheels/scipy/files

**Comment by lesteve** [comment]
Super nice, thanks for this! I will look at adding a similar build for scikit-learn then :wink:

**Comment by ngoldbaum** [comment]
> however that is either a numpy or a Python issue; it's visible in the numpy free-threaded wheel build [CI log](https://github.com/numpy/numpy/actions/runs/9377383356/job/25818853571) as well. @ngoldbaum is this a known thing?

It does show up after the call to print `np.show_config()` before the environment variable gets set, but it doesn't show up when the tests are executed:

```
  + export PYTHON_GIL=0
  + PYTHON_GIL=0
  + python -c 'import sys; import numpy; sys.exit(not numpy.test(label='\''full'\'', extra_argv=['\''-n=auto'\'']))'
  NumPy version 2.1.0.dev0+git20240604.24cdc31
  NumPy CPU features:  SSE SSE2 SSE3 SSSE3* SSE41* POPCNT* SSE42* AVX* F16C* FMA3* AVX2* AVX512F? AVX512CD? AVX512_KNL? AVX512_KNM? AVX512_SKX? AVX512_CLX? AVX512_CNL? AVX512_ICL?
  bringing up nodes...
  bringing up nodes...
  
  ........................................................................ [  0%]
  ........................................................................ [  0%]
  ........................................................................ [  0%]
```

I think the issue over here in scipy is that the environment variable is getting set in `cibw_before_test.sh`, but then `cibw_test_command.sh` doesn't see the variable set in the other script. NumPy doesn't use `cibw_before_test.sh`.

**Comment by rgommers** [comment]
Thanks Nathan, that would make sense.

> I think the issue over here in scipy is that the environment variable is getting set in `cibw_before_test.sh`, but then `cibw_test_command.sh` doesn't see the variable set in the other script. NumPy doesn't use `cibw_before_test.sh`.

@andfoy could you investigate and fix in a follow-up?

**Comment by andfoy** [comment]
That could be indeed the case here, let me fix it in a follow-up

**Pull request merged**